modules/globallock/lock.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package globallock
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"code.gitea.io/gitea/modules/nosql"
+	"code.gitea.io/gitea/modules/setting"
+
+	redsync "github.com/go-redsync/redsync/v4"
+	goredis "github.com/go-redsync/redsync/v4/redis/goredis/v9"
+)
+
+type Locker interface {
+	Lock() error
+	TryLock() (bool, error)
+	Unlock() (bool, error)
```

Comment:
I think it's necessary to clarify the behaviors of these methods for some edge cases.

I see differences between different implementations.

| Case | memoryLock | redisLockService | potential problem |
| - | - | - | - |
| `TryLock()` for a locked mutex | return `false, nil` | return `false, error` | `if ok, err := r.TryLock(); err != nil { return err  /*there's really no error here*/} else if ok //...` |
| `Unlock()` for a unlocked mutex | panic | return `false, nil` | The returned `bool, error` will confuse the developers that `Unlock` can be used as `TryUnlock` |

Maybe more. I know there could be some challenges to align memoryLock and redisLockService, but it's worth to be done.